### PR TITLE
36 unique reports - Closes #36

### LINKED ISSUE #36: Add unique IDs for unique reports

### PR DESCRIPTION
**Pull request by robertavram-md**
Fixes #36 
This pull request includes several changes across multiple files to improve the configuration, add new functionality, and enhance the testing suite. The key changes include updates to configuration files, improvements in the retrieval metrics computation, and the addition of new tests.

Configuration updates:

* [`config/base_config.yaml`](diffhunk://#diff-ad781b6517d6292f33686c44469e3576f928320a4b4996cd88fb4411f98e351bL11-R11): Updated `data_filename` to use a smaller sample dataset for training.
* `config/sweep_config_single_video.yaml` and `config/sweep_config_single_video.yaml.bak`: Modified parameters for learning rate, optimizer, scheduler, and batch size to support a more flexible and comprehensive sweep configuration. [[1]](diffhunk://#diff-736b873bc7ffd63be441a212ebee0c1a04046133a93c971b47bf7d25cb3254b6L3-R3) [[2]](diffhunk://#diff-736b873bc7ffd63be441a212ebee0c1a04046133a93c971b47bf7d25cb3254b6L22-R30) [[3]](diffhunk://#diff-736b873bc7ffd63be441a212ebee0c1a04046133a93c971b47bf7d25cb3254b6L37-R47) [[4]](diffhunk://#diff-736b873bc7ffd63be441a212ebee0c1a04046133a93c971b47bf7d25cb3254b6L83-R85)

Code improvements:

* [`models/text_encoder.py`](diffhunk://#diff-91e55674b1f986e241a4551f0d015cde06d21869cd02fb8b5a31655131b272d8R49-R55): Added tokenizer initialization in the `__init__` method to ensure the tokenizer is available for text encoding.
* [`runners/video_constrative_learning.py`](diffhunk://#diff-545e6c560da8c3adbc5af3516ae3d55a4aaa3115723e4d08bd5abc98b832db06L403-R478): Enhanced the `_run_epoch` method to compute NxM retrieval metrics, deduplicate texts, and save text embeddings for future use. Also removed unnecessary debug prints. [[1]](diffhunk://#diff-545e6c560da8c3adbc5af3516ae3d55a4aaa3115723e4d08bd5abc98b832db06L403-R478) [[2]](diffhunk://#diff-545e6c560da8c3adbc5af3516ae3d55a4aaa3115723e4d08bd5abc98b832db06L456-R512)

Testing enhancements:

* [`tests/test_retrieval_metrics.py`](diffhunk://#diff-7fd8c1e117a1259c27915f5061fdb06147fc855547282948dfc3a34beff75e6dR1-R157): Added comprehensive tests for retrieval metrics, including perfect retrieval, random retrieval, worst retrieval, non-square similarity matrices, and edge cases.

Metric computation updates:

* [`utils/metrics.py`](diffhunk://#diff-b4ad8049203cd9bbf0c3f3a4edb37f8b72780228334d55fab31c4c555f942bd6L164-R171): Fixed the return type of the `compute_median_rank` function to return an integer instead of a float. [[1]](diffhunk://#diff-b4ad8049203cd9bbf0c3f3a4edb37f8b72780228334d55fab31c4c555f942bd6L164-R171) [[2]](diffhunk://#diff-b4ad8049203cd9bbf0c3f3a4edb37f8b72780228334d55fab31c4c555f942bd6L186-R186)

## Summary by Sourcery

Improves the configuration, adds new functionality, and enhances the testing suite. The key changes include updates to configuration files, improvements in the retrieval metrics computation, and the addition of new tests.

Bug Fixes:
- Fixes the return type of the `compute_median_rank` function to return an integer instead of a float.
- Fixes issue #36

Enhancements:
- Enhances the `_run_epoch` method to compute NxM retrieval metrics, deduplicate texts, and save text embeddings for future use.
- Adds tokenizer initialization in the `__init__` method to ensure the tokenizer is available for text encoding.
- Updates `data_filename` to use a smaller sample dataset for training.
- Modifies parameters for learning rate, optimizer, scheduler, and batch size to support a more flexible and comprehensive sweep configuration.
- Saves unique texts and their indices.
- Saves text embeddings for future use

Tests:
- Adds comprehensive tests for retrieval metrics, including perfect retrieval, random retrieval, worst retrieval, non-square similarity matrices, and edge cases.

### CODE REVIEW

**Comment by robertavram-md** [comment]
See training run here https://wandb.ai/mhi_ai/dev_deep_coro_clip_single_video/sweeps/2wc463sa?nw=nwuserrobertavram

**Comment by JacquesDelfrate** [comment]
Looks good - je merge. On pourra les fix les units test dans un prochain PR. 
Je vais en ajouter pour le linear prob. dans un future issue.
